src/inspect_ai/model/_providers/goodfire.py

```diff
@@ -0,0 +1,286 @@
+
+        # Validate model name against supported models
+        supported_models = list(get_args(SUPPORTED_MODELS))
+        if self.model_name not in supported_models:
+            raise ValueError(f"Model {self.model_name} not supported. Supported models: {supported_models}")
+
+        # Initialize client with remaining model args
+        base_url_val = model_base_url(base_url, "GOODFIRE_BASE_URL")
+        assert isinstance(base_url_val, str) or base_url_val is None
+
+        self.client = GoodfireClient(
+            api_key=self.api_key,
+            base_url=base_url_val or DEFAULT_BASE_URL,
+            **self.model_args,
+        )
+
+        # Initialize variant with specified name if provided
+        variant_model = MODEL_MAP.get(self.model_name, "meta-llama/Meta-Llama-3.1-8B-Instruct")
+        if variant_model not in get_args(SUPPORTED_MODELS):
+            raise ValueError(f"Variant {variant_model} not supported. Supported variants: {list(get_args(SUPPORTED_MODELS))}")
+        # NOTE: There's a type mismatch between Goodfire's runtime behavior and type hints
+        # The docs show direct string usage: variant = goodfire.Variant("meta-llama/Meta-Llama-3-8B-Instruct")
+        # But the type hints expect a Literal. We validate the model name above, so this is safe at runtime.
+        # TODO: Consider creating an issue in Goodfire's repo about this type mismatch
+        self.variant = Variant(variant_model)  # type: ignore
+
+    def _to_goodfire_message(self, message: ChatMessage) -> GoodfireChatMessage:
+        """Convert an Inspect message to a Goodfire message format.
+        
+        Args:
+            message: The message to convert
+            
+        Returns:
+            The converted message in Goodfire format
+            
+        Raises:
+            ValueError: If the message type is unknown
+        """
+        role: Literal["system", "user", "assistant"] = "user"
+        if isinstance(message, ChatMessageSystem):
+            role = "system"
+        elif isinstance(message, ChatMessageUser):
+            role = "user"
+        elif isinstance(message, ChatMessageAssistant):
+            role = "assistant"
+        elif isinstance(message, ChatMessageTool):
+            role = "user"  # Convert tool messages to user messages
+        else:
+            raise ValueError(f"Unknown message type: {type(message)}")
+
+        content = str(message.content)
+        if isinstance(message, ChatMessageTool):
+            content = f"Tool {message.function}: {content}"
+
+        return cast(GoodfireChatMessage, {
+            "role": role,
+            "content": content,
+        })
+
+    def handle_error(self, ex: Exception) -> ModelOutput | Exception:
+        """Handle API errors and convert to appropriate outputs."""
+        error_msg = str(ex).lower()
+        
+        # Only handle rate limits and return other errors as-is
+        if self.is_rate_limit(ex):
```

Comment:
We don't want to handle rate limits here, rather we want these exception to propagate up so our higher level retry handler can catch them and retry.  The `handle_error()` used by other models is used almost exclusively for handling context window overflows and refusals (which are frustratingly reported via exception not in a structured way). This function exists essentially to take these conditions and return them into `ModelOutput` with `StopReason` of `model_length` or `content_filter`.